src/Verraes/Lambdalicious/datastructures.php

```diff
@@ -33,33 +35,82 @@ function ispair($pair)
 }
 
 /**
- * Get the first element off a list
+ * Get the first element of a list
  *
- * @param pair|array
- * @return mixed|array
+ * @param pair|list
+ * @return mixed|list
  */
 function head($data)
 {
-    return
-        // for compatibility with arrays (@TODO get rid of this later?)
-        is_array($data) ?
-            (isempty($data) ? raise("head() is only defined for non-empty lists.") : reset($data)) :
-            $data(head);
+    return $data(head);
 }
 
 /**
  * Returns the second element of the pair, or returns the list without its first element
  *
- * @param pair|array
- * @return mixed|array
+ * @param pair|list
+ * @return mixed|list
  */
 function tail($data)
 {
-    return
-        // for compatibility with arrays (@TODO get rid of this later?)
-        is_array($data) ?
-            (isempty($data) ? raise("tail() is only defined for non-empty lists.") : array_slice($data, 1)) :
-        $data(tail);
+    return $data(tail);
 }
 
 final class pair {} // for IDE's
+
+/**
+ * Creates a list
+ */
+function l(...$elements)
+{
+    $createList = function(array $elements, $list = 'λ_list') use (&$createList) {
+        if (empty($elements)) {
+            return $list;
+        }
+
+        $last = current(array_reverse($elements));
+
+        $newList = pair($last, $list);
+
+        return $createList(array_reverse(array_slice(array_reverse($elements), 1)), $newList);
+    };
+
+    return $createList($elements);
+}
```

Comment:
It can be done much simpler, like this

```php
/**
 * Creates a list
 */
function l(...$elements)
{
    $createList = function(array $elements, $list = 'λ_list') use (&$createList) {
        if (empty($elements)) {
            return $list;
        }

        $last = array_pop($elements);
        $newList = pair($last, $list);

        return $createList($elements, $newList);
    };

    return $createList($elements);
}
```

though, like this, we're mutating the `$elements` array using `array_pop()`, so maybe that's not the best way :confused: